### PR TITLE
MAILBOX-358 Get ride of store events

### PR DESCRIPTION
**Pull request by hoangdat**
- Introduce POJO implementation for MailboxEvent in MailboxListener

- Remove EventImpl in EventFactory

- Generating API events in EventFactory 

Jira: https://issues.apache.org/jira/projects/MAILBOX/issues/MAILBOX-358

### CODE REVIEW

File: mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java [hunk 1]
```diff
@@ -314,65 +324,96 @@ public MetaDataHoldingEvent(MailboxSession.SessionId sessionId, User user, Mailb
...
 
     /**
      * A mailbox event related to updated flags
      */
-    abstract class FlagsUpdated extends MessageEvent {
+     class FlagsUpdated extends MessageEvent {
 
         /**
          *
          */
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java]: Can you remove empty comments?

File: mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java [hunk 2]
```diff
@@ -314,65 +324,96 @@ public MetaDataHoldingEvent(MailboxSession.SessionId sessionId, User user, Mailb
...
         private static final long serialVersionUID = 1L;
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java]: Can we remove serialVersionUID ?

File: mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java [hunk 3]
```diff
@@ -314,65 +324,96 @@ public MetaDataHoldingEvent(MailboxSession.SessionId sessionId, User user, Mailb
...
 
-        public FlagsUpdated(MailboxSession.SessionId sessionId, User user, MailboxPath path, MailboxId mailboxId) {
+        private final List<MessageUid> uids;
+        private final List<UpdatedFlags> uFlags;
+
+        public FlagsUpdated(MailboxSession.SessionId sessionId, User user, MailboxPath path, MailboxId mailboxId, List<MessageUid> uids, List<UpdatedFlags> uFlags) {
             super(sessionId, user, path, mailboxId);
+            this.uids = ImmutableList.copyOf(uids);
+            this.uFlags = ImmutableList.copyOf(uFlags);
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java]: uFlags is IMO not a good name

File: protocols/imap/src/test/java/org/apache/james/imap/processor/base/FakeMailboxListenerAdded.java [hunk 4]
```diff
@@ -38,7 +38,7 @@
     public List<MessageUid> uids;
 
     public FakeMailboxListenerAdded(MailboxSession.SessionId sessionId, User user, List<MessageUid> uids, MailboxPath path, MailboxId mailboxId) {
-        super(sessionId, user, path, mailboxId);
+        super(sessionId, user, path, mailboxId, null);
```
Comment (chibenwa) [inline protocols/imap/src/test/java/org/apache/james/imap/processor/base/FakeMailboxListenerAdded.java]: What is `null`?

File: mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java [hunk 5]
```diff
@@ -314,65 +324,96 @@ public MetaDataHoldingEvent(MailboxSession.SessionId sessionId, User user, Mailb
...
+        private final Map<MessageUid, MessageMetaData> uids;
+
+        public Expunged(MailboxSession.SessionId sessionId, User user, MailboxPath path, MailboxId mailboxId, Map<MessageUid, MessageMetaData> uids) {
             super(sessionId, user, path, mailboxId);
+            this.uids = uids;
+        }
+
+        @Override
+        public List<MessageUid> getUids() {
+            return ImmutableList.copyOf(uids.keySet());
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java]: it's too late, if you think the uids are mutable, you should make a defensive copy in constructor

File: mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java [hunk 6]
```diff
@@ -314,65 +324,96 @@ public MetaDataHoldingEvent(MailboxSession.SessionId sessionId, User user, Mailb
...
          */
+        public MessageMetaData getMetaData(MessageUid uid) {
+            return added.get(uid);
+        }
+
         @Override
-        public abstract MessageMetaData getMetaData(MessageUid uid);
-        
+        public List<MessageUid> getUids() {
+            return ImmutableList.copyOf(added.keySet());
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java]: same comment about the place you do the copy

File: protocols/imap/src/test/java/org/apache/james/imap/processor/base/FakeMailboxListenerFlagsUpdate.java [hunk 7]
```diff
@@ -31,27 +31,11 @@
 
 public class FakeMailboxListenerFlagsUpdate extends FlagsUpdated {
```
Comment (mbaechler) [inline protocols/imap/src/test/java/org/apache/james/imap/processor/base/FakeMailboxListenerFlagsUpdate.java]: why do we need this class in the first place ?

Comment (hoangdat) [reply protocols/imap/src/test/java/org/apache/james/imap/processor/base/FakeMailboxListenerFlagsUpdate.java]: I removed this class

File: mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java (hunk 5 above)
Comment (hoangdat) [reply mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java]: I fixed it by defensive copy in constructor.

File: mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java (hunk 6 above)
Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java]: It is a conversion to `List` and not a copy for immutability

File: protocols/imap/src/test/java/org/apache/james/imap/processor/base/FakeMailboxListenerAdded.java [hunk 8]
```diff
@@ -33,12 +33,14 @@
 import org.apache.james.mailbox.model.MessageId;
```
Comment (chibenwa) [inline protocols/imap/src/test/java/org/apache/james/imap/processor/base/FakeMailboxListenerAdded.java]: Remove that useless class please.

Use directly `MailboxListener.Added` in `MailboxEventAnalyserTest`

File: mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java (hunk 5 above)
Comment (mbaechler) [reply mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java]: It's probably better to change the return type in MessageEvent to Set, Collection or Iterable

File: mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java (hunk 6 above)
Comment (mbaechler) [reply mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java]: could be changed to Set, Collection or Iterable in MessageEvent IMO

File: protocols/imap/src/test/java/org/apache/james/imap/processor/base/MailboxEventAnalyserTest.java [hunk 9]
```diff
@@ -134,33 +144,32 @@ public void testShouldBeNoSizeChangeOnOtherEvent() {
...
+        testee.event(MAILBOX_ADDED_EVENT);
         testee.resetEvents();
 
         assertThat(testee.isSizeChanged()).isFalse();
     }
 
     @Test
     public void testShouldNotSetUidWhenNoSystemFlagChange() {
-        FakeMailboxListenerFlagsUpdate update = new FakeMailboxListenerFlagsUpdate(MAILBOX_SESSION,
+        MailboxListener.FlagsUpdated update = new MailboxListener.FlagsUpdated(
```
Comment (mbaechler) [inline protocols/imap/src/test/java/org/apache/james/imap/processor/base/MailboxEventAnalyserTest.java]: why don't you use EventFactory ?

File: mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java (hunk 5 above)
Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java]: :+1: for Collection

**Comment by chibenwa** [comment]
```
[bf880de9140570d30e8dbe60090b3a90d22c273b] [ERROR] /james-project/mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerStressTest.java:[78,57] cannot find symbol
[bf880de9140570d30e8dbe60090b3a90d22c273b]   symbol:   method get(int)
[bf880de9140570d30e8dbe60090b3a90d22c273b]   location: interface java.util.Collection<org.apache.james.mailbox.MessageUid>
[bf880de9140570d30e8dbe60090b3a90d22c273b] [INFO] 1 error
```

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerStressTest.java [hunk 10]
```diff
@@ -75,8 +75,11 @@ public ListenerType getType() {
 
             @Override
             public void event(Event event) {
-                MessageUid u = ((Added) event).getUids().get(0);
-                uList.add(u);
+                Collection<MessageUid> uids = ((Added) event).getUids();
+                if (!uids.isEmpty()) {
```
Comment (mbaechler) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerStressTest.java]: you don't need the `if` as you know exactly the events that will be received

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**